core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -757,8 +753,11 @@ private[spark] class ExecutorAllocationManager(
       val taskIndex = taskEnd.taskInfo.index
       val stageId = taskEnd.stageId
       allocationManager.synchronized {
-        if (stageIdToNumRunningTask.contains(stageId)) {
-          stageIdToNumRunningTask(stageId) -= 1
+        if (taskIdToNumLiveTasks.contains(taskId)) {
+          taskIdToNumLiveTasks(taskId) -= 1
+          if (taskIdToNumLiveTasks(taskId) <= 0) {
```

Comment:
We shall still ensure it's not < 0 ?

Reply:
In theory, there is no negative number of tasks.
Just to prevent negative numbers and to clear them in time.

Reply:
In that case if we don't expect negative number of tasks to happen I would insert a assertion here, otherwise we can fail fast and print sufficient information.